### PR TITLE
[Fix] Form element status accessibility

### PR DESCRIPTION
**Pull request by LJKaski**
## Description
This PR makes status text an aria-live region, so screen reader users get information of status changes immediately. It also implements this change into form components using StatusText.

## Motivation and Context
The lack of information about component status changes was something that has come up in the recent accessibility clinics and needed to be addressed.

## How Has This Been Tested?
Tested with NVDA + chrome + FF and Talkback + Chrome.

### TextInput, TextArea, Checkbox, SearchInput
* Status text now has `aria-live="assertive"` by default. This can be changed via `statusTextAriaLiveMode` prop.

### CODE REVIEW

File: src/utils/aria.ts [hunk 1]
```diff
@@ -5,6 +5,8 @@ export const ariaLabelOrHidden = (ariaLabel?: string) =>
     ? { 'aria-label': ariaLabel, role: 'img' }
     : { 'aria-hidden': true };
 
+export type ariaLiveModes = 'assertive' | 'polite' | 'off';
```
Comment (aappoalander) [inline src/utils/aria.ts:8]: We should probably name types using these conventions: https://basarat.gitbook.io/typescript/styleguide#type

And should use PascalCase for these?

File: src/core/Form/StatusText/StatusText.test.tsx [hunk 2]
```diff
@@ -10,11 +10,6 @@ describe('props', () => {
       const { container } = render(<StatusText>Test text</StatusText>);
       expect(container.firstChild).toHaveTextContent('Test text');
     });
-
-    it('is null when no children given', () => {
```
Comment (aappoalander) [inline src/core/Form/StatusText/StatusText.test.tsx:14]: Should there still be a test to check that it renders correctly even without children?

File: src/core/Form/StatusText/StatusText.test.tsx [hunk 3]
```diff
@@ -43,13 +38,6 @@ describe('props', () => {
       expect(container.firstChild).toHaveClass('fi-status-text--error');
     });
   });
-
-  describe('disabled', () => {
-    it('is null when disabled', () => {
```
Comment (aappoalander) [inline src/core/Form/StatusText/StatusText.test.tsx:48]: Same here, maybe we should check that the aria live is not enabled when a component is disabled?

File: src/core/Form/StatusText/StatusText.tsx [hunk 4]
```diff
@@ -24,10 +24,17 @@ export interface StatusTextProps extends HtmlSpanProps {
...
 const StyledStatusText = styled(
-  ({ className, children, status, ...passProps }: StatusTextProps) => (
+  ({
+    className,
+    children,
+    disabled,
+    status,
+    ...passProps
+  }: StatusTextProps) => (
     <HtmlSpan
```
Comment (aappoalander) [inline src/core/Form/StatusText/StatusText.tsx:34]: I would check here if the disabled prop is true and remove aria-live if not needed.

File: src/utils/aria.ts (hunk 1 above)
Comment (ketsappi) [reply src/utils/aria.ts:8]: So the suggested type would be after changes like this, if I'm on the same line:
`export type AriaLiveMode = 'assertive' | 'polite' | 'off';`

- Using PascalCase
- No plural in name

File: src/core/Form/StatusText/StatusText.tsx [hunk 5]
```diff
@@ -24,10 +24,17 @@ export interface StatusTextProps extends HtmlSpanProps {
 }
```
Comment (ketsappi) [inline src/core/Form/StatusText/StatusText.tsx:25]: ### This might be over abstraction and not needed

As the same `statusTextAriaLiveMode` and `statusText` prop is repeated over and over what about if we change it to separate type?

```tsx
// src/core/Form/types.ts
export interface StatusText {
  /** Status text to be shown below the component and hint text. Use e.g. for validation error */
  statusText?: string;

  /** Aria-live mode for the status text element
   * @default assertive
   */
  statusTextAriaLiveMode?: ariaLiveModes;
}
```

Then just extending that type instead, `SearchInput` as an example:
```tsx
// src/core/Form/SearchInput/SearchInput.tsx
export interface SearchInputProps
  extends Omit<
      HtmlInputProps,
      | 'type'
      | 'disabled'
      | 'onChange'
      | 'children'
      | 'onClick'
      | 'value'
      | 'defaultValue'
    >,
    StatusText {
	//...
}
```

Comment (aappoalander) [reply src/core/Form/StatusText/StatusText.tsx:25]: I think that's a good idea. Assuming it fits all use cases, which in this case I think it does. Should not be too big of a task to change. But I think we still need to export the granular types like AriaLiveMode as well.

File: src/utils/aria.ts (hunk 1 above)
Comment (LJKaski) [reply src/utils/aria.ts]: Already changed it to PascalCase, but will remove the plural too.

File: src/core/Form/StatusText/StatusText.tsx (hunk 5 above)
Comment (LJKaski) [reply src/core/Form/StatusText/StatusText.tsx:29]: I like the idea as well. Maybe a thing for a separate ticket just to keep things clean on that front?

Comment (aappoalander) [reply src/core/Form/StatusText/StatusText.tsx:29]: Fits here as well in my opinion. Would save a round of testing without affecting issues scope much. But up to author to decide.

Comment (LJKaski) [reply src/core/Form/StatusText/StatusText.tsx:29]: That's fine as well :) I'll do it while I'm at it.

File: src/utils/aria.ts [hunk 6]
```diff
@@ -5,6 +5,8 @@ export const ariaLabelOrHidden = (ariaLabel?: string) =>
     ? { 'aria-label': ariaLabel, role: 'img' }
     : { 'aria-hidden': true };
 
+export type AriaLiveMode = 'assertive' | 'polite' | 'off';
+
```
Comment (ketsappi) [inline src/utils/aria.ts:9]: Nitpicking here, but as this is not an util it could be located to `src/core/Form/types.ts`

_On the other end, filename `types.ts` is also somehow misleading when the `interface StatusTextCommonProps` was added. But interface and type declarations seems to close in TypeScript so maybe we can live with this._

File: src/utils/aria.ts (hunk 1 above)
Comment (LJKaski) [reply src/utils/aria.ts]: Should now be in order.

File: src/core/Form/StatusText/StatusText.test.tsx (hunk 2 above)
Comment (LJKaski) [reply src/core/Form/StatusText/StatusText.test.tsx:14]: Added a test for this.

File: src/core/Form/StatusText/StatusText.test.tsx (hunk 3 above)
Comment (LJKaski) [reply src/core/Form/StatusText/StatusText.test.tsx:48]: This is now tested.

File: src/core/Form/StatusText/StatusText.tsx (hunk 4 above)
Comment (LJKaski) [reply src/core/Form/StatusText/StatusText.tsx]: This should now work as described.

File: src/core/Form/StatusText/StatusText.tsx (hunk 5 above)
Comment (LJKaski) [reply src/core/Form/StatusText/StatusText.tsx:29]: Implemented as `StatusTextCommonProps` interface.

File: src/utils/aria.ts (hunk 6 above)
Comment (LJKaski) [reply src/utils/aria.ts]: Moved the type to types.ts

File: src/core/Form/StatusText/StatusText.test.tsx (hunk 3 above)
Comment (aappoalander) [reply src/core/Form/StatusText/StatusText.test.tsx:48]: I'm wondering how we should handle this. It might be safer to keep aria-live attribute in DOM always when status text component is used. When there is no content, live mode should be off and same goes with disabled. Possible values are off, polite or assertive. This test should also check that the attribute exists rather than just checking there is children present.

File: src/core/Form/StatusText/StatusText.tsx (hunk 4 above)
Comment (aappoalander) [reply src/core/Form/StatusText/StatusText.tsx]: Reconsidering this to use `aria-live="off"` only when set with props or when component is disabled. Otherwise it should be always on, even when there is no content.

File: src/core/Form/StatusText/StatusText.test.tsx (hunk 2 above)
Comment (aappoalander) [reply src/core/Form/StatusText/StatusText.test.tsx:14]: See other comments for better explanation, but I think we should always keep aria-live attribute in DOM.

File: src/core/Form/StatusText/StatusText.tsx (hunk 4 above)
Comment (LJKaski) [reply src/core/Form/StatusText/StatusText.tsx]: That might be a better option. Not sure how differently the approaches would work in practice, but this one should be a cleaner solution at least 👍

File: src/core/Form/StatusText/StatusText.test.tsx (hunk 3 above)
Comment (LJKaski) [reply src/core/Form/StatusText/StatusText.test.tsx:48]: When there's no content, I don't see why it should be off. The way I think about it is that having an empty region with assertive aria-live mode means that it is ready to read out loud whatever content it gets. Turning it off while waiting for content doesn't really accomplish anything in my opinion. And for the test, do you mean we should add a test to check if an empty element has the aria-live attribute as well?

File: src/core/Form/StatusText/StatusText.tsx (hunk 4 above)
Comment (LJKaski) [reply src/core/Form/StatusText/StatusText.tsx]: It should now be 'off' when the element is disabled and on with the given mode in other cases - even when empty.

File: src/core/Form/StatusText/StatusText.test.tsx (hunk 3 above)
Comment (LJKaski) [reply src/core/Form/StatusText/StatusText.test.tsx:48]: Added tests to see that the element has an aria-live attribute even when empty and that the mode is 'off' when the element is disabled.

File: src/core/Form/StatusText/StatusText.test.tsx (hunk 2 above)
Comment (LJKaski) [reply src/core/Form/StatusText/StatusText.test.tsx:14]: Commented elsewhere. Should now be fixed.

**Review by ketsappi — APPROVED** [review]

**Review by ketsappi — APPROVED** [review]
Tested again after the rebase. Seems to work as intended.
Small hiccups with the Brave browser with aria-live as discussed elsewhere but we are not able to effect that.

**Review by aappoalander — APPROVED** [review]
